### DISCUSSION TITLE
Using `ConformalIntervals` with `make_reduction` on hierarchical datasets

### DISCUSSION THREAD

**Discussion by yarnabrina** [alan-turing-institute/sktime#5089] [Q&A]
## Reproducible Example

```python
from sklearn.linear_model import LinearRegression

from sktime.forecasting.compose import ForecastX, make_reduction
from sktime.forecasting.conformal import ConformalIntervals
from sktime.forecasting.model_selection import temporal_train_test_split
from sktime.forecasting.var import VAR
from sktime.utils._testing.hierarchical import _make_hierarchical

df = _make_hierarchical(
    hierarchy_levels=(2, 3),
    max_timepoints=90,
    min_timepoints=90,
    n_columns=4,
    all_positive=True,
    random_state=0,
)

y = df[["c0", "c1"]]
X = df[["c2", "c3"]]

y_train, y_test, X_train, X_test = temporal_train_test_split(y, X=X, test_size=15)

regressor = LinearRegression()
endogenous_model = make_reduction(regressor, pooling="global")  # local fails option 2

exogenous_model = VAR()

try:
    forecaster = ConformalIntervals(ForecastX(endogenous_model.clone(), exogenous_model.clone()))

    forecaster.fit(y_train, X=X_train, fh=range(1, 3))  # fails (No valid mtype)

    forecaster.predict(X=X_test)
    forecaster.predict_interval(X=X_test)
except Exception as error:
    print(f"option 1 fails: {error}")
else:
    print("option 1 works")

try:
    forecaster = ForecastX(ConformalIntervals(endogenous_model.clone()), exogenous_model.clone())

    forecaster.fit(y_train, X=X_train, fh=range(1, 3))  # fails (No valid mtype)

    forecaster.predict(X=X_test)
    forecaster.predict_interval(X=X_test)
except Exception as error:
    print(f"option 2 fails: {error}")
else:
    print("option 2 works")

try:
    forecaster = ForecastX(endogenous_model.clone(), exogenous_model.clone())

    forecaster.fit(y_train, X=X_train, fh=range(1, 3))

    forecaster.predict(X=X_test)
    # forecaster.predict_interval(X=X_test)  # not supported
except Exception as error:
    print(f"option 1 without conformal fails: {error}")
else:
    print("option 1 without conformal works")

try:
    forecaster = ForecastX(endogenous_model.clone(), exogenous_model.clone())

    forecaster.fit(y_train, X=X_train, fh=range(1, 3))

    forecaster.predict(X=X_test)
    # forecaster.predict_interval(X=X_test)  # not supported
except Exception as error:
    print(f"option 2 without conformal fails: {error}")
else:
    print("option 2 without conformal works")
```

## Results

```pycon
❯ python conformal_test.py
option 1 fails: No valid mtype could be identified for object of type <class 'pandas.core.frame.DataFrame'>. Errors returned are as follows, in format [mtype]: [error message] 
pd.DataFrame: <class 'pandas.core.indexes.multi.MultiIndex'> is not supported for obj, use one of (<class 'pandas.core.indexes.range.RangeIndex'>, <class 'pandas.core.indexes.period.PeriodIndex'>, <class 'pandas.core.indexes.datetimes.DatetimeIndex'>) or integer index instead.
pd.Series: obj must be a pandas.Series, found <class 'pandas.core.frame.DataFrame'>
np.ndarray: obj must be a numpy.ndarray, found <class 'pandas.core.frame.DataFrame'>
xr.DataArray: obj must be a xarray.DataArray, found <class 'pandas.core.frame.DataFrame'>

option 2 fails: No valid mtype could be identified for object of type <class 'pandas.core.frame.DataFrame'>. Errors returned are as follows, in format [mtype]: [error message] 
pd.DataFrame: <class 'pandas.core.indexes.multi.MultiIndex'> is not supported for obj, use one of (<class 'pandas.core.indexes.range.RangeIndex'>, <class 'pandas.core.indexes.period.PeriodIndex'>, <class 'pandas.core.indexes.datetimes.DatetimeIndex'>) or integer index instead.
pd.Series: obj must be a pandas.Series, found <class 'pandas.core.frame.DataFrame'>
np.ndarray: obj must be a numpy.ndarray, found <class 'pandas.core.frame.DataFrame'>
xr.DataArray: obj must be a xarray.DataArray, found <class 'pandas.core.frame.DataFrame'>

option 1 without conformal works
option 2 without conformal works
```

## Version

`main` branch, commit 08e2d662f

**Comment by yarnabrina** [comment]
Additional side question: if I use `local` instead of `global`, `"option 2"` fails with this:

> The `window_length` and `fh` are incompatible with the length of `y`

But since each series has 90 observations, `window_length` in `make_reduction` is `10` because of default value and `fh` is at most `2`, I do not understand what is the problem. I understood `local` creates one model per series (lowest hierarchy) and global creates only one across all series, and neither case could cause `window_length` + `fh_max` >= `n_timepoints` as far as I can think, so I'm very confused.

FYI: @hazrulakmal because you explained this very well in #4753.

**Reply by yarnabrina** [reply]
Hi @hazrulakmal, sorry to tag you once again, but just making sure in case you missed this side question.

**Reply by hazrulakmal** [reply]
Thank you for tagging me again. I apologise for overlooking this earlier. I believe there may be an issue with the inner workings of the `ConformalIntervals` estimator that is causing this error. While I am not familiar with the inner workings of conformal intervals beyond knowing that they can help produce confidence interval predictions for arbitrary estimators, this seems to be a likely culprit.

reason being, If you fit the `make_reduction` estimator only without conformal interval, the said error is not raised

```python
from sklearn.linear_model import LinearRegression

from sktime.forecasting.compose import  make_reduction
from sktime.forecasting.model_selection import temporal_train_test_split
from sktime.utils._testing.hierarchical import _make_hierarchical

df = _make_hierarchical(
    hierarchy_levels=(2, 3),
    max_timepoints=90,
    min_timepoints=90,
    n_columns=4,
    all_positive=True,
    random_state=0,
)

y = df[["c0", "c1"]]
X = df[["c2", "c3"]]

y_train, y_test, X_train, X_test = temporal_train_test_split(y, X=X, test_size=15)

regressor = LinearRegression()
endogenous_model = make_reduction(regressor, pooling="local")

forecaster = endogenous_model.clone()
forecaster.fit(y_train, X=X_train, fh=range(1, 3)) 
forecaster.predict(X=X_test)
```

I suggest maybe there is need to do some debugging and keeping tracking of the variable `window_length` `fh_max` & `n_timepoints` to actually see what the conformal interval does to these variable.

**Reply by fkiraly** [reply]
apologies for overlooking the side question.

The reason for the behaviour is that `ConformalIntervals` itself does a sliding window inside to estimate the prediction intervals, although this is hard-coded as full sliding window, and parameterized by `initial_window`, and `sample_frac`.

The default `initial_window` is `max(10, 0.1*len(y))`, which is 10 for your `y`, so less than the 12 a full fit of `make_reduction` would need.
It should work if you make `initial_window` in `ConformalIntervals` larger.

**Reply by fkiraly** [reply]
@hazrulakmal, indeed your guess was right!
Maybe we should clarify the windowing properties in the docstring, instead of just deferring to the literature reference.

**Comment by fkiraly** [comment]
This failure is due to a bug that snuck in when we extended the forecasters to hierarchical.

I opened a bug report issue in #5092.

The `ConformalIntervals` wrapper would clone the `X_inner_mtype` and `y_inner_mtype` tags of the inner forecaster.
If that tag was hierarchical, it would pass a hierarchical object to `_fit`, which it did not support - it supports only `Series` scitype.

The bug did not occur and was not caught in the tests, since it happens rarely that estimator tags `X_inner_mtype` or `y_inner_mtype` include a hierarchical mtype - and that is precisely the condition to surface the bug.

This has been fixed by just passing all `Series` mtypes through, and not providing any panel/hierarchical ones, thus prompting vectorization.
See https://github.com/sktime/sktime/pull/5091

**Reply by fkiraly** [reply]
Update: #5091 was not the full fix - see https://github.com/sktime/sktime/pull/5093 for the fix in the case of hierarchical wrapped estimator.
